### ISSUE TITLE
Query for global cursor position

### ISSUE THREAD

**Issue by adambrown** [glfw/glfw#922]
Sometimes it's necessary to have absolute mouse coordinates, and GLFW does not seem to provide any way to get this currently. Being able to call `glfwGetCursorPos()` without a window handle or with a 0 window handle and getting back the absolute mouse coordinates in the virtual desktop if the cursor is enabled would be ideal.

Reasons for this are sometimes you need to know where the mouse is to know where you want to create a new window. If you are building a floaty window style app like old school Photoshop or Gimp you need to know what the mouse is doing relative to the whole space.

Dealing with Windows 10 "warp lines" (see https://github.com/glfw/glfw/issues/921 ) requires knowing when the mouse has crossed one and the window relative positions don't take this into account in the current screen coordinate system.

**Comment by elmindreda** [comment]
Passing `NULL` for the window seems like a good solution. Not sure yet if this will make it for 3.3. Also need to look into Wayland and Mir.

**Comment by jadahl** [comment]
This API will not be portable. Some display architectures (Wayland, Mir at least) does not and will not support querying the global cursor positioning.

**Comment by adambrown** [comment]
Is it necessary for a feature to be supported on all target platforms to make it into GLFW? I've only been using GLFW for a few months, so not 100% familiar with the project's policy and politics. Up until now I've been hacking around missing functionality in GLFW by using AWT to fill in everything I need to support multiple scale factor monitors, among other things, on Windows 10. I'm using GLFW, OpenGL and NanoVG through LWJGL. However, I now need to get the application working on OSX and it's not possible to use AWT and GLFW in the same application on OSX. You have to use the `-XstartOnFirstThread` flag and can not touch any classes within AWT. So to this end, I've started adding feature requests to GLFW to fill in the missing functionality I've been using AWT to work around.

Getting the global cursor position is possible on the bigger 2 target platforms (by number of users for our case) via win forms and cocoa. It seems to me a large segment of commercial applications may care more about these two platforms.

It would be a shame if features that can support 2 of the 3 platforms couldn't make it in because the third didn't have it. It seems that if most platforms support a feature but one doesn't, a simple note, about this feature is not supported on platform _____, would do the trick, and let developers choose whether they care about that or if they need to work around it.

Not that linux support isn't great, but it's just so much work for some smaller teams to worry about only 2 platforms that are relatively stable and consistent. We couldn't even consider supporting a 3rd platform that comes in a ton of different flavors and fully testing on all of them for a comparatively micro user segment. ...Oh if only we had enough time, money and developers to implement every high priority feature AND still have cycles left over to focus on a 3rd platform! That's the dream. To be honest I'm questioning whether we can even support 2. It might be necessary to release the product first on Windows only, but I'd like to give OSX a fair shot first, and maybe come back to it if the product is a success.

**Comment by elmindreda** [comment]
It can be implemented on X11 as well, which is still the dominant display server on Linux and other Free Unices. Window positions also are similarly not available to clients on Wayland and Mir, despite the GLFW API supporting it, so adding this would not be unprecedented.

**Comment by jadahl** [comment]
Adding API for things like multi monitor scaling makes complete sense; and support for it being poor on some platforms (e.g. X11) should indeed not stop the feature to be introduced, but things that doesn't actually *need* API and are actually used as hacks to implement some actual features (for example global pointer positions have traditionally been used as a hack to implement pointer locking) just makes it harder to use GLFW to implement portable applications. It's not unlikely that future platforms will stop allowing hacks involving global positioning as it effectively makes it impossible to implement certain types of features in the display architecture.

**Comment by adambrown** [comment]
Fair enough. I agree with your sentiment. I'm definitely using global mouse position to implement "hacks" that could be built-in features.

For example, being able to drag an un-decorated window between screens is really only possible to do smoothly and robustly using global mouse position. When you try and implement it by getting the mouse position relative to the thing you are dragging, which I tried first, it comes out very jittery and unnatural. Switching to remembering the global mouse start position at the beginning of a drag operation and not starting the drag until the mouse had moved by some factor (10sc worked well for me), then calculating the window position based on the delta movement from drag start position cleared it right up and made it smooth and natural.

GLFW could definitely add an `setDragWindowRect(Rect DragArea)` type thing and implement all of this under the hood. That would be awesome. However, this is one of a bunch of such "hacks". I'd have to request features for every such thing I want to add. In my experience it's usually harder to get features like that added to libraries like GLFW and takes a lot longer for everyone involved to agree on the right behavior and API, so I opted to ask for the simpler solution which gets me a win faster by allowing me to implement these "hacks" with a smaller amount of work by a GLFW dev.

I don't think the two approaches are mutually exclusive though. As GLFW continues to grow and add new features, and new platforms that don't support some of these hacks come out and provide alternatives to the hacks, so GLFW is forced to implement a lot more high level functionality, users like me can stop using this low level functionality, but until then, if access to the low level functionality exists, we can all use it to be productive and develop cool stuff in the mean time.

**Comment by jadahl** [comment]
IMHO it should be those types of features (for example dragging an undecorated window would be a one-liner in the Wayland backend) that a toolkit abstracts, since these are the things that will work very differently on different platforms. By forcing applications to implement half of the abstraction them self (like in the window-drag case), which a toolkit aimed to do for the application, the toolkit failed to abstract the platform. It is true, though, indeed, at one point a feature becomes a hack. I'm sure some abstraction libraries back in the days e.g. had a way to directly access framebuffer memory; an abstraction might have been sane when it was introduced but eventually would become obsolete, but adding *new* API to add things that are not necessary and doomed to sooner rather than later become obsolete, seems backward to me.

**Comment by adambrown** [comment]
So if everyone seems to think implementing this functionality myself is a step backwards, then should I close this and ask for addition of a drag window area instead? The other hacks I'm using this for are to fix issues caused by #676, #902 and #898 and I expect to not need them once multiple non-100% dpi scales are supported on windows 10 properly, and undecorated windows are properly resizable using the native window functionality.

I am doing exactly what @Andos is doing here: #902. Creating windows with completely re-purposed title bar area. The current trend on Windows 10, and possibly some other platforms as well, seems to be that the title bar is dead (Firefox, Chrome, Edge, Photoshop, Visual Studio, and a multitude of others, too numerous to name or remember off the top of my head, seem to agree). This space can be better used by moving up your menu, toolbar, tabs or whatever you have right below the title bar in order to maximize usable space. These apps all still provide some portion of the former title bar as a drag area for moving the window however.

The last remaining hack that I need the functionality requested in this issue is for implementing this drag area.

**Comment by elmindreda** [comment]
Custom window hit-testing is a feature I've considered adding, again modulo Wayland and Mir.

**Comment by adambrown** [comment]
closing this in favor of: https://github.com/glfw/glfw/issues/923

**Comment by jadahl** [comment]
What is "custom hit testing"?

**Comment by zao** [comment]
@jadahl Some window platforms have native functionality for your application to indicate that parts of the window should be considered special. Among those is Windows' [WM_NCHITTEST](https://msdn.microsoft.com/en-us/library/windows/desktop/ms645618(v=vs.85).aspx) which lets you consider parts of the window a caption bar for native dragging purposes. Similar messages/events assuredly exist for other platforms.

**Comment by jadahl** [comment]
@zao So it is something meant to make it possible to have custom window header bar etc, ala Chrome? If so, this works slightly different on Wayland. There the client draws the window decorations, so things like a maximize button, move or resize area are already "custom" to the client. Though GLFW doesn't support drawing any decorations under Wayland/Mir yet. Anyhow, if this is what hit testing is about I suspect it'd be trivial to implement under Wayland (and possibly Mir too, I'm not sure) as long as the GLFW API is done with this in mind.

**Comment by adambrown** [comment]
@jadahl yes that's exactly what it is for. I closed this, since getting the global cursor position was more of a catchall for reimplementing a number of features, most of which will be addressed by other issues: #676, #902, #898. The only outstanding hack not covered by those other issues was #923, so I made a much more specific ticket for that piece of functionality and dropped this catchall since it seems it's not needed, and it's slightly against the style of GLFW.
